src/util.rs

```diff
@@ -1,32 +1,118 @@
-use chrono::offset::Utc;
-use chrono::DateTime;
 use futures_io::AsyncRead;
 use futures_util::ready;
 use ring::digest;
 use std::io::{self, ErrorKind};
 use std::marker::Unpin;
 use std::pin::Pin;
 use std::task::{Context, Poll};
+use std::time::{Duration, Instant};
 
 use crate::crypto::{HashAlgorithm, HashValue};
 use crate::Result;
 
+pub(crate) trait SafeAsyncRead: Sized {
```

Comment:
It's probably fine that this is `pub(crate)` for now, but I can imagine custom `Repository` implementations might want to use this functionality.